django/db/models/fields/__init__.py

```diff
@@ -1722,6 +1724,8 @@ def to_python(self, value):
         return decimal_value
 
     def get_db_prep_save(self, value, connection):
+        if hasattr(value, "as_sql"):
+            return value
         return connection.ops.adapt_decimalfield_value(
```

Comment:
If we were to move `adapt_decimalfield_value` into `get_db_prep_value` we should be able to get rid of `get_db_prep_save` here completely? Or is there a reason for doing this only on save

Reply:
Opened as https://github.com/django/django/pull/16264 -- let's see

Reply:
Allowing `get_db_prep_value` to adapt, given it's also used when a value is passed to lookups, will likely beneficial for  #15687 if it was having issues with the lack of `::numeric` casting when strings were passed to `filter(numeric__field=13.37)` or `filter(numeric__field="13.37")`.

In short, anything that makes more adequate types (e.g. `Jsonb` instead of a `str`) consistently and coherently make their way to the underlying backend empowers it to take the right decision with regards to typing.